Display scaling, live window resizing, and high-DPI support

past-due opened the pull request:
Trac: [#4720](http://developer.wz2100.net/ticket/4720)

### Overview

1. Internal enhancements to support screen / game / rendering scaling factors.
2. Internal enhancements to support live window resizing / resolution changes.
3. Internal enhancements to support **UI layout recalculation**.
4. **Support for live window resizing / resolution changes.**<sup>[SDL backend]</sup>
5. **"Display Scale"** option in the Video options menu, offering 100%+ display scaling options. <sup>[SDL backend]</sup>
6. Automatic high-DPI ("retina") display support. <sup>[macOS-only]</sup>
7. Better caching of text rasterization and reduction in unnecessary OpenGL state changes (required to support higher display scaling levels).
8. Better clipping and other small tweaks (required to support higher display scaling levels).

-----

### General

These changes all essentially depend on each other, thus this is a single PR.

### Additional Details

1. Text and other rendering now supports scaling factors.
2. New functions in all of the various UI-related components / files enable a call to a new parent `gameScreenSizeDidChange` function that distributes notifications of screen size changes.
3. Several enhancements of note:
    - `WIDGET` now supports a new `calcLayout` callback function / lambda that is called when layout needs to be recalculated.
    - `WIDGET` now supports a new `onDelete` callback function / lambda that is called when the instance is about to be deleted.
    - `W_INIT` now supports a new `initPUserDataFunc` attribute which can be used to initialize the `pUserData` pointer per widget instance.
    - Parent `WIDGET`s automatically distribute screen size change notifications to all children.
    - Where needed, UI widgets have been migrated to setting a `calcLayout` lambda.
4. Simply open the **Video options** and change the Resolution. **Or resize the window at the corners / edges** (as supported by your OS).
    > Current Limitations:
    > - Only live resolution changes **in the current mode** are supported. If you change the display mode (windowed / fullscreen), changes will be queued until Warzone is restarted (matching previous behavior). This is due to several current SDL quirks.
    > - **Live resizing is disabled while in a multiplayer game** because pausing is disabled in multiplayer games (and with how the current event loop + SDL works, the game effectively pauses while the window is being actively resized).
    > - Only the SDL backend currently supports live resizing. Compiling with the QT backend will revert to the prior behavior.
5. Open the **Video options** and select from the new **Display Scale** percentages (100%+) available at the current window size. (Display scale options are limited by the window / screen size, as no combination can result in a logical screen size lower than the minimum supported by Warzone. Hence, only users with very large displays - and the Warzone window expanded on those displays - will see the highest display scaling options. Most users probably won't see over 250% available.)
    > Current Limitations:
    > - Only the SDL backend currently supports Display Scale options. Compiling with the QT backend will not currently show the Display Scale option in the menu.
6. On macOS, automatic support for high-DPI ("retina") displays is enabled. Once a future version of SDL enables proper support for other OSes, they can be supported.
7. Some highlights:
    - Switched over most text rendering to using the improved `WzText`, which caches and reduces text rasterization calls by over 99%, for a significant performance win. (Previously, most text was being re-rasterized on every frame, even if unchanged.)
    - A lot of time was previously being wasted disabling shaders... in loops that just kept enabling / disabling the same shader. New functions to batch calls and moving `pie_DeactivateShader` outside of loops yield a significant performance improvement (especially when running at higher display scaling modes that necessarily stress the engine more).
8. Previously, the game engine spent a lot of time rendering things that were far off the visible screen. More accurate clipping techniques yield a large reduction in wasted non-visible drawing.

Forgon2100 commented:
Performance must be significantly improved before this patch can be accepted.

This should be discussed in ticket #4720 and the forums, so more people can help.

Chewbakka-Wakka commented:
Fantastic if it has. Perhaps we can test before and after patch on the performance difference. Got any volunteers? I would if I get time.

Sent from my iPhone

> On 27 Jan 2018, at 11:15, Forgon2100 <notifications@github.com> wrote:
> 
> Performance must be significantly improved before this patch can be accepted.
> 
> This should be discussed in ticket #4720 and the forums, so more people can help.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

swiftlyfalling commented:
@Forgon2100: As mentioned on Trac, all my tests on several Windows and Mac machines showed improved performance, so we'll need to track down exactly what's happening on your system.

perim commented:
This functionality is really nice! Thanks.

1. This patch adds requirement to SDL2 2.0.4+, since it uses SDL_GetDisplayDPI, which was only added in this version. Please change the version check in configure.ac to match.
2. Gigantic commits are difficult to test and nearly always have bugs that are then difficult to bisect. Please at least split the last two points (general optimizations) into a separate commit.
3. src/intdisplay.cpp:1228:18: error: variable ‘iDX’ set but not used

past-due commented:
@perim:
1. Made the SDL 2.0.4+ change to `configure.ac`
2. Split into two commits, per your request.
3. Fixed. 
The revised patch should also fix the performance regression @Forgon2100 experienced.

past-due commented:
@perim: 

I force pushed a fix to the branch. `configure.ac` actually needs a minimum version of 2.0.5 (because of `SDL_SetWindowResizable`).

past-due commented:
@Forgon2100: The above (3rd) commit should fix the issue you encountered on ArchLinux. (I reproduced it on Ubuntu.)

An SDL quirk meant that `SDL_GetWindowSize` returned cached values (which do *not* match the adjusted window size that the desktop window manager adopts when attempting to set a window size greater than the **usable screen space** - i.e. the screen minus system reserved areas). This did not occur on macOS or Windows. To avoid this, the code now obtains the usable screen space with `SDL_GetDisplayUsableBounds` and rejects attempts to set the window size to something greater than permitted.

past-due commented:
@per: So I dug into this a bit more, and there's still an issue caused by a bug in SDL (not picking up "window manager resized the window" events) and attempting to set the windowed window size via the Video options menu. (Testing with i3 window manager on Ubuntu).

I'm not aware of an easy workaround for this SDL bug. Since this set of patches enables OS/window-manager provided window resizing (dragging by corners, etc), I'd like to suggest the following change we spoke briefly about on IRC:
- Disabling the "Resolution" option from the Video options menu when in *windowed* mode.

It just isn't necessary anymore - users can just resize the window normally with these patches - and it apparently can't be made to work right in all circumstances because of SDL bugs.

DroidSparks commented:
warzone2100\lib\widget\widget.cpp(223): fatal error C1017: invalid integer constant expression

Use #ifdef DEBUG to fix it.

past-due commented:
Merged in: 619e84d56189f45138ff18b3aea733dc15d9ca0c, 4d0dcae247c6d93b884adfbfc0046017b66d6abd, e58b34a0ed5ca670bb8107343f8a01b96a5cfc59, 96b4866269ecd2c88bb260e282df830598642a0f
